### PR TITLE
Handle float16/bfloat16 to uint16 properly in helper functions

### PR DESCRIPTION
**Pull request by jcwchen**
**Description**
- Add conversion float16->uint16 in make_tensor so users can simply give it a float list or numpy array
- Do the same thing for bloat16
- Add tests

**Motivation and Context**
https://github.com/onnx/onnx/issues/3715 Currently users need additional code to convert float16 tensor. `make_tensor` should handle the conversion to simplify the usage.

### CODE REVIEW

File: onnx/helper.py [hunk 1]
```diff
@@ -240,15 +240,24 @@ def make_tensor(
         assert not raw, "Can not use raw_data to store string type"
 
     # Check number of vals specified equals tensor size
-    size = 1 if (not raw) else (mapping.TENSOR_TYPE_TO_NP_TYPE[data_type].itemsize)
+    expected_size = 1 if (not raw) else (mapping.TENSOR_TYPE_TO_NP_TYPE[data_type].itemsize)
+    # Flatten a numpy array if its rank > 1
+    if type(vals) is np.ndarray and len(vals.shape) > 1:
```
Comment (askhade) [inline onnx/helper.py:245]: what is the current behavior for ndarrays?

File: onnx/numpy_helper.py [hunk 2]
```diff
@@ -57,23 +57,23 @@ def to_array(tensor):  # type: (TensorProto) -> np.ndarray[Any]
         if (tensor_dtype == TensorProto.COMPLEX64
                 or tensor_dtype == TensorProto.COMPLEX128):
             data = combine_pairs_to_complex(data)
-        # F16 is stored as int32; Need view to get the original value
-        if tensor_dtype == TensorProto.FLOAT16:
+        # float16/bfloat16 is stored as int32 (uint16 type); Need view to get the original value
```
Comment (askhade) [inline onnx/numpy_helper.py:60]: nit: structure this a little different 
```
        # float16/bfloat16 is stored as int32 (uint16 type); Need view to get the original value
        if (tensor_dtype == TensorProto.FLOAT16
                or tensor_dtype == TensorProto.BFLOAT16):
            return (
                np.asarray(
                    tensor.int32_data,
                    dtype=np.uint16)
                .reshape(dims)
                .view(np.float16))
        data = getattr(tensor, storage_field),  # type: Sequence[np.complex64]
        if (tensor_dtype == TensorProto.COMPLEX64 
          or tensor_dtype == TensorProto.COMPLEX128):
            data = combine_pairs_to_complex(data)
   
        return (
            np.asarray(
                data,
                dtype=storage_np_dtype)
            .astype(np_dtype)
            .reshape(dims)
        )
```

File: onnx/mapping.py [hunk 3]
```diff
@@ -19,6 +19,7 @@
     int(TensorProto.INT64): np.dtype('int64'),
     int(TensorProto.BOOL): np.dtype('bool'),
     int(TensorProto.FLOAT16): np.dtype('float16'),
+    int(TensorProto.BFLOAT16): np.dtype('float16'),
```
Comment (askhade) [inline onnx/mapping.py:22]: add a comment explaining why you are mapping bfloat16 to np float16 something like (numpy does not support bfloat16...)

File: onnx/numpy_helper.py [hunk 4]
```diff
@@ -59,27 +59,27 @@ def to_array(tensor, base_dir=""):  # type: (TensorProto, Text) -> np.ndarray[An
...
-        # Otherwise simply use astype to convert; e.g., int->float, float->float
-        else:
-            return (
-                np.asarray(
-                    data,
-                    dtype=storage_np_dtype)
-                .astype(np_dtype)
-                .reshape(dims)
-            )
+        data = getattr(tensor, storage_field),  # type: Sequence[np.complex64]
```
[the remark this answers is not in the document]
Comment (jcwchen) [reply onnx/numpy_helper.py]: Yes it is incorrect... Just removed it. Thanks

File: onnx/test/helper_test.py [hunk 5]
```diff
@@ -391,6 +391,30 @@ def test_make_tensor(self):  # type: () -> None
         )
         self.assertEqual(string_list, list(tensor.string_data))
 
+    def test_make_float16_tensor(self):  # type: () -> None
```
Comment (askhade) [inline onnx/test/helper_test.py:394]: can you add a test with raw=True for both FLOAT16 and BFLOAT16

Comment (jcwchen) [reply onnx/test/helper_test.py:394]: Added. To make it work, I made split_complex_to_pairs and numpy_array.view go into the else condition (raw=False). IIUC, the vals for raw=True should be the final binary data after conversion so no more conversion is needed inside this helper function (make_tensor).

**Review by askhade — APPROVED** [review]

**Pull request merged**